Reward payments 1

ethanfrey opened the pull request:
Recreate from #69 after a mistaken merge and revert.

Enables rewards to be sent every epoch from consumer to producer and distributed to stakers.
Stakers can withdraw rewards to an account on the consumer chain (where they can be staked), with a message on the provider chain (which knows who staked)

maurolacy reviewed (approved):
Looks good! Will take a look at tests later.

ethanfrey commented on packages/apis/src/ibc/packet.rs:
```diff
@@ -56,6 +69,13 @@ pub enum ConsumerPacket {
     /// It contains a list of `valoper_address` to be removed
     RemoveValidators(Vec<String>),
+    /// This is part of the rewards protocol
+    Distribute {
```
That is a very good question. I made #81 to track this later. Not sure what the proper answer is right now. Not blocking merge but good to keep track of it.

ethanfrey commented on docs/ibc/Rewards.md:
```diff
@@ -133,3 +133,23 @@ packets over multiple channels.
+tokens in reserve (always in the case of a correct provider chain), it must release
+them to the given address. The consumer chain must be designed to handle rollbacks here
+on any error (especially due to invalid recipient address that cannot be validated on the
+provider side).
```
Exactly. I mean ICS20 never moves funds either. But it prints a transferable voucher on the recipient chain. Here we don't allow the vouchers to be trader, just redeemed directly on the consumer chain.

ethanfrey commented:
@maurolacy I addressed your comments in https://github.com/osmosis-labs/mesh-security/pull/76/commits/0a7f5635412b22b2e30120f6474cee46ec3dd0f5 if you want to check. And then rebased on main and resolved conflicts with other merges. I think this is more or less ready now.

maurolacy commented:
> @maurolacy I addressed your comments in [0a7f563](https://github.com/osmosis-labs/mesh-security/commit/0a7f5635412b22b2e30120f6474cee46ec3dd0f5) if you want to check. And then rebased on main and resolved conflicts with other merges. I think this is more or less ready now.

Will review tests, just for completeness.

maurolacy commented on contracts/provider/external-staking/src/multitest.rs:
```diff
@@ -316,11 +316,7 @@ fn staking() {
-        Tx::InFlightRemoteStaking { id, .. } => *id,
-        Tx::InFlightRemoteUnstaking { id, .. } => *id,
-    })
+    txs.first().map(Tx::id)
```
Nice.

maurolacy commented on contracts/provider/external-staking/src/multitest.rs:
```diff
@@ -755,16 +752,14 @@ fn distribution() {
     contract
-        .distribute_rewards(validators[0].to_owned())
-        .with_funds(&coins(50, STAR))
+        .test_distribute_rewards(validators[0].to_owned(), coin(50, STAR))
```
This is better / fixed now. Rewards to distribute are virtual here.

This `test_` prefix is a little ugly, but makes things clear. Hopefully it will be gone when we're able to test this better.

ethanfrey replied:
Yes, when both Sylvia and cw-multi-test have IBC support this should be much better.
Until then we just fake all ibc packet lifetimes.

I wouldn't assume this to improve til later Q3 (it is on Q3 roadmap though)

maurolacy commented on contracts/provider/external-staking/src/multitest.rs:
```diff
@@ -805,29 +800,28 @@ fn distribution() {
-            .unwrap(),
-        coins(80, STAR)
-    );
+    // assert_eq!(
+    //     app.app()
+    //         .wrap()
+    //         .query_all_balances(contract.contract_addr.clone())
+    //         .unwrap(),
+    //     coins(80, STAR)
+    // );
```
Remove commented code.

maurolacy reviewed (commented):
Some comments after a quick review on tests.
